### ISSUE TITLE
implement Meyers diffing algorithm

### ISSUE THREAD

**Issue by yoshuawuyts** [choojs/nanomorph#85]
So we have a slight diffing problem — not only are we fairly slow at updating, reordering nodes is also suboptimal. This means that some reorders can be surprising at the worst case, and in a worst case reorders simply take longer than they should.

In computer science this is known as "diffing", and is a rather well-studied problem. Probably the best known algo for this is the Mayers Diffing algorithm. It'd be neat if we could make use of this for `nanomorph`.

### references
- https://github.com/yelouafi/petit-dom/blob/master/src/vdom.js
- https://neil.fraser.name/writing/diff/
- http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.4.6927&rep=rep1&type=pdf
- https://www.codeproject.com/Articles/42279/Investigating-Myers-diff-algorithm-Part-of

**Comment by yoshuawuyts** [comment]
@graforlock did you get a chance to look at this?

**Comment by graforlock** [comment]
Yeah @yoshuawuyts made some steps to extract it along with pre-optimisations that can be applied to nanomorph, but the newborn has beaten me off the track :) Will probably progress with it this weekend

**Comment by blahah** [comment]
Myers diffing for the dom is totally tractable. @graforlock looking forward to your implementation.

**Comment by graforlock** [comment]
Problems might be with numbers of pre-optimisations and we need quite a bit of them. DOM node getters and setters should be used sparsely as they are quite slow in bulk, and when comparing real DOM trees this needs special attention as any use of aforementioned slows the computation down.

**Comment by blahah** [comment]
@graforlock I've been thinking and reading about the problem and I actually think Myers is unlikely to be the best approach. It's good for the general diff problem, but for a DOM tree there are specialised algorithms.

<img width="1070" alt="screen shot 2017-10-23 at 00 03 38" src="https://user-images.githubusercontent.com/836040/31867549-04bc08ae-b792-11e7-9382-3ca48b68d970.png">

http://referaat.cs.utwente.nl/conference/3/paper/7144/change-detection-in-xml-trees-a-survey.pdf

The DOM is a tree with ordered siblings, and we want the operations:

- insert
- delete
- update
- move

So from the list above, the [DiffXML](http://diffxml.sourceforge.net/) might be worth looking at.

**Comment by graforlock** [comment]
Myers is a groundwork for most of the other algorithms including tree based diffs. Its also proven to do good work in many libraries that operate on trees. Yes, it is not specialised but there are specialised and well-optimized versions at hand.

**Comment by kristoferjoseph** [comment]
I say try them all, but ship as iterations pass the test suite.
Good news is we can try out different approaches in branches now that we have a baseline and make incremental improvements.

**Comment by blahah** [comment]
@graforlock sounds good to me 👌

**Comment by graforlock** [comment]
@blahah or anyone, I struggle to find a whole lot of time at the moment, so I'm progressing slowly. Although I have good suites and understanding how to make it work well, we could pair up on that can't we?

**Comment by yoshuawuyts** [comment]
@graforlock maybe if you could create a repo, people could like dig in and help out? I'd love for this to make some progress — there's not a lot of pressure, but still, the sooner we'd get a version working the better :sparkles:

**Comment by blahah** [comment]
yeah, a repo for a pure `myers-diff-dom` or something would be ideal, then we can collab in there?

**Comment by graforlock** [comment]
yeah will do, a good idea

**Comment by yoshuawuyts** [comment]
@graforlock yay! — looking forward to it! :sparkles:

**Comment by graforlock** [comment]
should do tomorrow, need to polish some edges and README this stuff

**Comment by graforlock** [comment]
So, heres the repo @yoshuawuyts @blahah @kristoferjoseph 

It is `petit-dom` algorithm ported/adapted to a DOM diffing methodology. A lot of stuff is just opinionated result of the port (like input updates etc), but the algorithm itself is extractable to the separate module. Sorry its all in one file atm! Has to be changed definitely.

It may be that I wasted huge amount of my and your time and its all pile of crap!

Or not...  😄 Preliminary results give me a huge performance advantage over nanomorph (as you can see in example)

https://github.com/graforlock/diffing-implementation

**Comment by graforlock** [comment]
This kind of needs refining & _tailoring_ for nanomorph, if anyone else is interested then we can start woking on it on a nanomorph branch? Anyone?

**Comment by graforlock** [comment]
So my thoughts are, after dealing with this stuff for some time.

Firstly, traversing the actual DOM tree for a tree diff is costly.

Myers is super slow (with all the diff traceback it does by default) and won't perform well over certain operations cap. `petit-dom` is bypassing it by switching to a generic LCS algorithm, which is faster, but in nanomorph's case still too slow. Only makes performance difference with equal number of nodes, other than that, seems to have no edge over what nanomorph is currently doing.

There is a need for super performant and greedy LCS algortihm. I myself found a possible candidate but this is not that easy on the actual DOM tree. Proxy nodes is another thing to consider -they have to be keyed somewhere in a map given they would change their position in the DOM tree, remaining the same in the end.

**Comment by jongacnik** [comment]
hey all! popped up in my dash today so just dropping this in, in case it is at all relevant → https://github.com/WebReflection/domdiff

**Comment by graforlock** [comment]
Looks really cool, how's performance of it ? Did you try to measure it?

**Comment by jongacnik** [comment]
@graforlock not yet, just a quick glance over source. Have a feeling you, @blahah, or @yoshuawuyts would be able to discern a bit more than myself!

**Comment by soyuka** [comment]
Good discussion about diffing: https://github.com/snabbdom/snabbdom/issues/348 
Think it's worth mentioning here.

**Comment by WebReflection** [comment]
not sure I'm adding anything to this discussion but for what I can tell with my tests, the petit-dom algo seems to be the fastest but there are too many operations that non virtual DOM wouldn't need, and compared to snabbdom is quite bigger, in terms of final size ( domdiff is 524 bytes in gzip and 471 bytes via brotli, as example ).

domdiff also wouldn't care less about patches and it works linearly with a list of nodes, it's not suitable for nested nodes too, just lists of the same parent node.

**Comment by olstenlarck** [comment]
Heya! :tada: Following this long time. And as i once mentioned, i once wanted to rewrite it from scratch - for learning, playing and minimizing the size. So code name was "minmorph" - removed the "./events" list, the remaining is almost the same with few optimizations. Actually not sure why we need to list all the events?

Long story short, back then i tested it against nanomorph's tests and only 1 tests was failing. Recently was touching it more and following the need for benchmarking, i got the @graforlock repo and put it there so we can benchmark. Also added the `nanomorph`s tests.

And so, the result is all of the _normal_ tests are passing and only 6 of the "keyed" tests are failing!
The benchmark results are **phenomenal** :)


**The tests are green (almost)**
![2018-02-02-02 34 40_1280x1024_scrot](https://user-images.githubusercontent.com/5038030/35710696-84e90fc0-07c1-11e8-8128-4ce139c243a1.png)

**minmorph benchmark** - numbers are the same as the @graforlock's not finished code, which in turn fails pretty much tests.
![2018-02-02-02 35 04_1280x1024_scrot](https://user-images.githubusercontent.com/5038030/35710694-84a6a25c-07c1-11e8-82d1-e2a90523faa6.png)

**Nanomorph benchmark**
![2018-02-02-02 34 53_1280x1024_scrot](https://user-images.githubusercontent.com/5038030/35710695-84cb9e0e-07c1-11e8-9e87-47c61f1fdec7.png)

**Comment by olstenlarck** [comment]
So, that's it https://github.com/tunnckoCore/demo-minmorph

```
yarn install
yarn dev
```

And open http://localhost:5000. It uses parcel-bundler to run on the browser btoh the tests and benchmarks.
Tests are at the `src/app/test.js` and not in the `test/` at the root. For trying the tests against `nanomorph` go to the file and pass `nanomorph` to the `abstractMorph` and `abstractKeyed` instead of `minmorph`.

**Comment by graforlock** [comment]
Cool stuff, I am going to have a look. 

I myself also experimented with "mostly" accurate LCS algorithm (Sift4), and honestly I don't think we  need proper 100% accurate diffing algorithms with backtrace provided if we have keyed results reordered properly. Sift seemed to be blazing fast and I may put some more work in it for greater good.

**Comment by olstenlarck** [comment]
Heya all! Just for the record of my work. Last night i started playing again on the above mentioned repo `demo-minmorph`. It's now a lot more organized and cleaned (reorganized and apply DRY on tests for #87), and deps are updated.
Please check it out and try running it.

Tests are increased as number, failing tests are decreased to only 2 (5 assertions).
Why tests are increased? They now are 724, because i switched to #104, now they are around 98 events (on chrome) instead of 30-35 hardcoded.

Only 8 of them are failing and are excluded from the suite

```js
function excludeEvents(name) {
  // failing event tests for Minmorph
  // Nanomorph fails for another ones and some of these
  const isFailing = [
    'onresume',
    'onfreeze',
    'onreadystatechange',
    'onpointerlockchange',
    'onpointerlockerror',
    'onselectionchange',
    'onvisibilitychange',
    'onsecuritypolicyviolation',
  ].includes(name);

  return !isFailing;
}
```

Most of the keyed tests are passing, except 2 (`handle non-id elements` and `nested without id`) which i marked as `TODO: PRs welcome` for easier search.

Benchmarks (`yarn bench`) are against Nanomorph@6 beta using the @graforlock's benchmark from ... hm, from where i got it (some repo). It's not something so special, but it shows that my implementation using a bit changed `domdiff` is around 3-5 times faster than Nanomorph. I'm still not sure for that numbers, because of the failing tests. 

Please share some feedback. I'll try to PR things one by one.

Cheers :clinking_glasses:

**Comment by brucou** [comment]
Hi guys,

what is the status on that?

**Comment by yoshuawuyts** [comment]
@brucou hey! -- I've casually started working on a WASM version, but there's no ETA when this might be done.

**Comment by brucou** [comment]
Alright, thanks for the info!

On 1/16/19, Yoshua Wuyts <notifications@github.com> wrote:
> @brucou hey! -- I've casually started working on a WASM version, but there's
> no ETA when this might be done.
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/choojs/nanomorph/issues/85#issuecomment-454764559


-- 
---

**Comment by tunnckoCore** [comment]
@brucou the demo-minmorph is there, you can see performance and issues there ;] I didn't looked on it for a while.

**Comment by JDvorak** [comment]
Any updates on this front?
